In [1]:
import numpy as np

The observational and model prediction data

In [2]:
observations = np.array([15.2, 16.1, 14.5, 15.8, 25.0])
model_output = np.array([14.8, 15.5, 14.0, 16.0, 25.3])

In [3]:
# Defining the covariance matrices
obs_cov = np.diag([0.1, 0.1, 0.1, 0.1, 0.1])  # Observational error covariance
model_cov = np.diag([0.1, 0.1, 0.1, 0.1, 0.1])  # Model error covariance

In [4]:
# Initializing the analysis field
analysis = np.zeros_like(model_output)

# Perform the Kalman filter data assimilation

In [5]:

for i in range(len(model_output)):
    # Computing the innovation
    innovation = observations[i] - model_output[i]
    
    # Computing the Kalman gain
    kalman_gain = model_cov[i, i] / (model_cov[i, i] + obs_cov[i, i])
    
    # Updating the analysis field
    analysis[i] = model_output[i] + kalman_gain * innovation



In [6]:
# Calculating the RMSE and bias
rmse_analysis = np.sqrt(np.mean((analysis - observations) ** 2))
bias_analysis = np.mean(analysis - observations)

# Print the results
print("Analysis field:", analysis)
print("RMSE (analysis vs. observations):", rmse_analysis)
print("Bias (analysis vs. observations):", bias_analysis)


Analysis field: [15.   15.8  14.25 15.9  25.15]
RMSE (analysis vs. observations): 0.2121320343559641
Bias (analysis vs. observations): -0.10000000000000035


Analysis field: [15.0, 15.8, 14.25, 15.9, 25.15]
Original observational data: [15.2, 16.1, 14.5, 15.8, 25.0]
NWP model output: [14.8, 15.5, 14.0, 16.0, 25.3]

Comparison:
The analysis field is closer to the observational data than the NWP model output.
The analysis field has a smaller range (15.0-25.15) compared to the NWP model output (14.0-25.3).
The analysis field has a smaller RMSE (0.212) and bias (-0.1) compared to the NWP model output (RMSE: 0.424, bias: -0.2).
The data assimilation process using the Kalman filter has improved the accuracy of the analysis field compared to the NWP model output.
The analysis field is closer to the observational data, with a smaller RMSE and bias.

In [7]:


# Calculate RMSE and bias for NWP model output vs. observations
rmse_nwp = np.sqrt(np.mean((model_output - observations) ** 2))
bias_nwp = np.mean(model_output - observations)

# Calculate RMSE and bias for analysis field vs. observations
rmse_analysis = np.sqrt(np.mean((analysis - observations) ** 2))
bias_analysis = np.mean(analysis - observations)

# Print evaluation results
print("RMSE (NWP model output vs. observations):", rmse_nwp)
print("Bias (NWP model output vs. observations):", bias_nwp)
print("RMSE (analysis field vs. observations):", rmse_analysis)
print("Bias (analysis field vs. observations):", bias_analysis)

# Evaluating improvements
if rmse_analysis < rmse_nwp:
    print("RMSE improved by", (rmse_nwp - rmse_analysis) / rmse_nwp * 100, "%")
else:
    print("RMSE degraded by", (rmse_analysis - rmse_nwp) / rmse_nwp * 100, "%")

if abs(bias_analysis) < abs(bias_nwp):
    print("Bias improved by", (abs(bias_nwp) - abs(bias_analysis)) / abs(bias_nwp) * 100, "%")
else:
    print("Bias degraded by", (abs(bias_analysis) - abs(bias_nwp)) / abs(bias_nwp) * 100, "%")


RMSE (NWP model output vs. observations): 0.4242640687119287
Bias (NWP model output vs. observations): -0.2
RMSE (analysis field vs. observations): 0.2121320343559641
Bias (analysis field vs. observations): -0.10000000000000035
RMSE improved by 50.00000000000006 %
Bias improved by 49.99999999999983 %




For RMSE (Root Mean Square Error):

RMSE (NWP model output vs. observations): 0.424
RMSE (analysis field vs. observations): 0.212

The RMSE decreased from 0.424 to 0.212, indicating a significant improvement in the accuracy of the analysis field compared to the NWP model output. This means that the data assimilation process reduced the error in the predictions by about 50% (0.212 / 0.424).
For Bias:

Bias (NWP model output vs. observations): -0.2
Bias (analysis field vs. observations): -0.1

The bias decreased from -0.2 to -0.1, indicating a reduction in the systematic error in the predictions. The bias is now closer to zero, indicating that the analysis field is less biased compared to the NWP model output.

Overall, the data assimilation process improved the accuracy and reduced the bias in the predictions, resulting in a more reliable analysis field. The RMSE and bias metrics provide a quantitative measure of the improvements made by the data assimilation process.

